x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/RepositoryAnalyzeAction.java

```diff
@@ -444,12 +448,29 @@ private boolean isRunning() {
             return true;
         }
 
+        private class CheckForCancelListener implements ActionListener<Void> {
+            @Override
+            public void onResponse(Void unused) {
+                // task complete, nothing to do
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                // trigger another isRunning check which will cancel the task if not already failed or cancelled
+                var isNowRunning = isRunning();
+                assert isNowRunning == false;
+            }
+        }
+
         public void run() {
             assert queue.isEmpty() : "must only run action once";
             assert failure.get() == null : "must only run action once";
 
             logger.info("running analysis of repository [{}] using path [{}]", request.getRepositoryName(), blobPath);
 
+            cancellationListener.addTimeout(request.getTimeout(), repository.threadPool(), EsExecutors.DIRECT_EXECUTOR_SERVICE);
```

Comment:
If the cluster has many nodes and the repo analysis is configured to have high concurrency, would it be expensive to cancel the tasks on the scheduler thread?

Reply:
Eh perhaps, but I wouldn't expect it to be a problem because (a) I don't see folks changing the concurrency very much and (b) even at 1000 nodes I don't think it'd be a huge deal, the cancel messages are tiny. We cancel things for other reasons on low-latency threads, e.g. `RestCancellableNodeClient`.